provider/aws.go

```diff
@@ -169,20 +200,54 @@ func (p *AWSProvider) submitChanges(zone string, changes []*route53.Change) erro
 		return nil
 	}
 
-	params := &route53.ChangeResourceRecordSetsInput{
-		HostedZoneId: aws.String(expandedHostedZoneID(zone)),
-		ChangeBatch: &route53.ChangeBatch{
-			Changes: changes,
-		},
+	zones, err := p.Zones()
+	if err != nil {
+		return err
 	}
 
-	if _, err := p.Client.ChangeResourceRecordSets(params); err != nil {
-		return err
+	// separate into per-zone change sets to be passed to the API.
+	changesByZone := changesByZone(zones, changes)
+
+	for z, cs := range changesByZone {
+		params := &route53.ChangeResourceRecordSetsInput{
+			HostedZoneId: aws.String(z),
+			ChangeBatch: &route53.ChangeBatch{
+				Changes: cs,
+			},
+		}
+
+		if _, err := p.Client.ChangeResourceRecordSets(params); err != nil {
+			return err
+		}
 	}
 
 	return nil
 }
 
+// changesByZone separates a multi-zone change into a single change per zone.
+func changesByZone(zones map[string]*route53.HostedZone, changeSet []*route53.Change) map[string][]*route53.Change {
+	changes := make(map[string][]*route53.Change)
+
+	for _, z := range zones {
```

Comment:
I'm not entirely sure. I would now add the "advanced" zone detection that tolerates multiple matching zones and then detects the best one based on number of subdomains (correct if that doesn't make sense).

I like the idea that endpoints that exist carry their hosted zone. This would take preference similar to the `RecordType` in https://github.com/linki/external-dns/blob/2ef0a608e1a7a798604d9aa623efe70c8e48dbcb/provider/aws.go#L228. However, this can be added in a second step.